pkg/mermaid/mermaidsequencediagram.go

```diff
@@ -0,0 +1,157 @@
+			retep := c.Ret.Payload
+			result += retStatement(appname, retep, previousapp, indent, thestart)
+		case *sysl.Statement_Action:
+			result += actionStatement(appname, c.Action.Action, indent)
+		case *sysl.Statement_Cond:
+			result += fmt.Sprintf("%salt %s\n", addIndent(indent), c.Cond.Test)
+			result += printSequenceDiagramStatements(m, c.Cond.Stmt, appname, previousapp, indent+1, sequencepairs, thestart)
+			if count+1 < len(statements) {
+				switch temp := statements[count+1].Stmt.(type) {
+				case *sysl.Statement_Group:
+					ok := startElse.MatchString(temp.Group.Title)
+					if !ok {
+						result += fmt.Sprintf("%send\n", addIndent(indent))
+					}
+				default:
+					result += fmt.Sprintf("%send\n", addIndent(indent))
+				}
+			} else {
+				result += fmt.Sprintf("%send\n", addIndent(indent))
+			}
+		case *sysl.Statement_Foreach:
+			result += fmt.Sprintf("%sloop %s\n", addIndent(indent), c.Foreach.Collection)
+			result += printSequenceDiagramStatements(m, c.Foreach.Stmt, appname, previousapp, indent+1, sequencepairs, thestart)
+			result += fmt.Sprintf("%send\n", addIndent(indent))
+		case *sysl.Statement_Loop:
+			result += fmt.Sprintf("%sloop %s\n", addIndent(indent), c.Loop.Criterion)
+			result += printSequenceDiagramStatements(m, c.Loop.Stmt, appname, previousapp, indent+1, sequencepairs, thestart)
+			result += fmt.Sprintf("%send\n", addIndent(indent))
+		case *sysl.Statement_LoopN:
+			result += fmt.Sprintf("%sloop %d times\n", addIndent(indent), c.LoopN.Count)
+			result += printSequenceDiagramStatements(m, c.LoopN.Stmt, appname, previousapp, indent+1, sequencepairs, thestart)
+			result += fmt.Sprintf("%send\n", addIndent(indent))
+		default:
+			panic("Unrecognised statement type")
+		}
+		count++
+	}
+	return result
+}
+
+func validAppNameAndEndpoint(m *sysl.Module, appname string, epname string) error {
+	if _, ok := m.Apps[appname]; !ok {
+		return errors.New("invalid app name")
+	}
+	if _, ok := m.Apps[appname].Endpoints[epname]; !ok {
+		return errors.New("invalid endpoint")
+	}
+	return nil
+}
+
+func callStatement(appname string, epname string, nextapp string, indent int) string {
+	var out = addIndent(indent)
+	out += fmt.Sprintf("%s->>+%s: %s\n", appname, nextapp, epname)
+	return out
+}
+
+func retStatement(appname string, epname string, previousapp string, indent int, thestart bool) string {
+	var out = addIndent(indent)
+	if thestart {
+		out += fmt.Sprintf("%s-->>%s: %s\n", appname, previousapp, epname)
+	} else {
+		out += fmt.Sprintf("%s-->>-%s: %s\n", appname, previousapp, epname)
+	}
+	return out
+}
+
+func actionStatement(appname string, action string, indent int) string {
+	var out = addIndent(indent)
+	out += fmt.Sprintf("%s->>%s: %s\n", appname, appname, action)
+	return out
+}
+
+func addIndent(indent int) string {
+	var out string
+	for i := 0; i < indent; i++ {
+		out += " "
+	}
+	return out
+}
+
+func sequencenpairscontain(s []sequencepair, sp sequencepair) bool {
```

Comment:
I think you can use map[string]struct{}{} and used `_, ok := myvar["foo"]` instead of storing in slices and then iteratng over slices

Reply:
Oh no can't do this as there will be multiple elements with the same key. The pair is what that makes sense, individual elements don't make much sense here.

Reply:
I can club all the strings together and have them as a key but is that necessary?